src/gui/editors/PianoRoll.cpp

```diff
@@ -2800,353 +2655,337 @@ void PianoRoll::paintEvent(QPaintEvent * pe )
-			// two white keys (e.g. between E and F)
-			++white_cnt;
-			if( white_cnt > 1 )
-			{
-				y -= m_whiteKeyBigHeight/2;
-			}
-		}
-
-		++key;
-	}
-
-
-	// erase the area below the piano, because there might be keys that
-	// should be only half-visible
-	p.fillRect( QRect( 0, key_line_y,
-			WHITE_KEY_WIDTH, noteEditBottom() - key_line_y ), bgColor );
-
-	// display note editing info
-	QFont f = p.font();
-	f.setBold( false );
-	p.setFont( pointSize<10>( f ) );
-	p.setPen( noteModeColor() );
-	p.drawText( QRect( 0, key_line_y,
-					  WHITE_KEY_WIDTH, noteEditBottom() - key_line_y ),
-			   Qt::AlignCenter | Qt::TextWordWrap,
-			   m_nemStr.at( m_noteEditMode ) + ":" );
-
-	// set clipping area, because we are not allowed to paint over
-	// keyboard...
-	p.setClipRect( WHITE_KEY_WIDTH, PR_TOP_MARGIN,
-				width() - WHITE_KEY_WIDTH,
-				height() - PR_TOP_MARGIN - PR_BOTTOM_MARGIN );
-
-	// draw the grid
-	if( hasValidPattern() )
-	{
-		int q, x, tick;
-
-		if( m_zoomingModel.value() > 3 )
+			if (key % KeysPerOctave == Key_C) { p.setPen(m_beatLineColor); }
+			else { p.setPen(m_lineColor); }
+			p.drawLine(m_whiteKeyWidth, y, width(), y);
+		};
+		// correct y offset of the top key
+		switch (prKeyOrder[topNote])
 		{
-			// If we're over 100% zoom, we allow all quantization level grids
-			q = quantization();
-		}
-		else if( quantization() % 3 != 0 )
-		{
-			// If we're under 100% zoom, we allow quantization grid up to 1/24 for triplets
-			// to ensure a dense doesn't fill out the background
-			q = quantization() < 8 ? 8 : quantization();
-		}
-		else {
-			// If we're under 100% zoom, we allow quantization grid up to 1/32 for normal notes
-			q = quantization() < 6 ? 6 : quantization();
-		}
-
-		// First we draw the vertical quantization lines
-		for( tick = m_currentPosition - m_currentPosition % q, x = xCoordOfTick( tick );
-			x <= width(); tick += q, x = xCoordOfTick( tick ) )
-		{
-			p.setPen( lineColor() );
-			p.drawLine( x, PR_TOP_MARGIN, x, height() - PR_BOTTOM_MARGIN );
+		case PR_WHITE_KEY_SMALL:
+		case PR_WHITE_KEY_BIG:
+			break;
+		case PR_BLACK_KEY:
+			// draw extra white key
+			drawKey(topKey + 1, grid_line_y - m_keyLineHeight);
 		}
-
-		// Draw horizontal lines
-		key = m_startKey;
-		for( int y = key_line_y - 1; y > PR_TOP_MARGIN;
-				y -= m_keyLineHeight )
+		// loop through visible keys
+		const int lastKey = qMax(0, topKey - m_pianoKeysVisible);
+		for (int key = topKey; key > lastKey; --key)
 		{
-			if( static_cast<Keys>( key % KeysPerOctave ) == Key_C )
+			bool whiteKey = Piano::isWhiteKey(key);
+			if (whiteKey)
 			{
-				// C note gets accented
-				p.setPen( beatLineColor() );
+				drawKey(key, grid_line_y);
+				drawHorizontalLine(key, grid_line_y);
+				grid_line_y += m_keyLineHeight;
 			}
 			else
 			{
-				p.setPen( lineColor() );
+				// draw next white key
+				drawKey(key - 1, grid_line_y + m_keyLineHeight);
+				drawHorizontalLine(key - 1, grid_line_y + m_keyLineHeight);
```

Comment:
If `lastKey + 1` is a white key, the loop will stop after drawing it, whereas it will draw `lastKey` if `lastKey + 1` is a black key. Is this the intended behavior?

Reply:
This handles the case where we draw a black key over a white key (current behavior). So, yes it's intended for that purpose.